### DISCUSSION TITLE
What do you think of nodes' placement in Children-Nodes Layout LEFTTORIGHT_BOTHSIDES_RIGHT?

### DISCUSSION THREAD

**Discussion by macmarrum** [freeplane/freeplane#1095] [Ideas]
This is how LEFTTORIGHT_BOTHSIDES_RIGHT looks like in v1.11.1-pre29

(1)
![LEFTTORIGHT_BOTHSIDES_RIGHT](https://user-images.githubusercontent.com/7197131/232134761-ed690943-02af-4976-a1a2-ee7c1f6b5610.png)

Let's look at the corresponding Layout of TOPTOBOTTOM_BOTHSIDES_BOTTOM. 

(2)
![TOPTOBOTTOM_BOTHSIDES_BOTTOM](https://user-images.githubusercontent.com/7197131/232133738-7b935017-2587-4412-b8e1-71939326aad6.png)

Here, the level-1 nodes are placed below Root, whereas in LEFTTORIGHT_BOTHSIDES_RIGHT, level-1 nodes are placed alongside Root.

This is how it would look like, if LEFTTORIGHT_BOTHSIDES_RIGHT would mimic TOPTOBOTTOM_BOTHSIDES_BOTTOM.

(3)
![LEFTTORIGHT_BOTHSIDES_RIGHT-proposed](https://user-images.githubusercontent.com/7197131/232134779-e16ba661-f0c6-42b7-bfc6-12425657493a.png)

I prefer this (3) over the original (1).

What do you tihnk?

**Comment by dpolivaev** [comment]
I agree that (3) looks more clear than (1). Should I change it?

**Reply by dpolivaev** [reply]
Changed in preview 1.11.1-pre30

**Comment by sifran-github** [comment]
Can you include a map?
I dont see the layout that you describe at (1) as extremely.
I took your map from https://github.com/freeplane/freeplane/issues/1084#issuecomment-1500952598, changed the layout of the root to 
![image](https://user-images.githubusercontent.com/5730355/232164823-a1d6ce41-d80f-4718-879a-770ea00d9541.png) (LEFTTORIGHT_BOTHSIDES_RIGHT).
and this is what i get in pre29:
![image](https://user-images.githubusercontent.com/5730355/232164887-aa2ab0bd-33ba-4df2-971b-105d037f50e8.png)

**Reply by macmarrum** [reply]
[childNodesLayout-POC.mm.zip](https://github.com/freeplane/freeplane/files/11239956/childNodesLayout-POC.mm.zip)

Please note that you can use the root-node scripts
- to recreate the initial state
- to switch to the next layout combination

**Reply by sifran-github** [reply]
OK, i had to play with the map and minimize the children, but now i see the problem.
It is happening when the parent is very large, and the children are very small. This is why:

![image](https://user-images.githubusercontent.com/5730355/232264188-f6b699e9-0072-4dd9-9b85-248e704dd91d.png)

in both cases the direction is to the RIGHT, but:
in the first example, it "looks like" the position of the children is to the right.
in the second example, it "looks like" the position of the children is to the left (because they are on the left side of their parent), which is not as expected.

I have some ideas how to solve it... i will either write a discussion about it, or talk privately.

**Reply by sifran-github** [reply]
Part of my idea is basically to allow the user more control over where exactly the child's starting position will be.
How do you do it? This is the question. If you've played with CSS design, you know that it can be a bit tricky.

for example:

Sometimes it is nice to have layouts where the children start from the LEFT side of the parent (and sometimes with a gap).
Sometimes it is nice to have layouts where the children start from the RIGHT side of the parent (and sometimes with a gap).
Sometimes it is nice to have layouts where the children start from the CENTER of the parent (and sometimes with a gap).

The tricky part is that there are many variables:

- What side of the parent do i want the child to be attached to? (left? center? right? maybe 30%? 50%+gap?)
- What side of the child do i want to start from there? (child's left? right? etc...)
- Gaps...
- (Maybe more variables are needed.)

(This example is for horizontal.  It should be for vertical decisions as well.)

**Reply by sifran-github** [reply]
The current solution that is implemented is to have some "predefined" layouts. It has many layouts which in most cases are great, but in other "edge cases" might look differently than expected.

if you break down the layouts into more basic properties and allow the user to edit them, it will allow more control over the layouts.
but the first issue to tackle is to come up with the exact PROPERTIES that define a layout... I spent a lot of time thinking about it, and I'm still not 100% sure. Anyway the current properties do allow some great layouts but not all the possible layouts (like in this case, it works great until the parent is very large).

and after we do define these properties, I'm not sure how it might look in the gui, because adding many more parameters to the format panel might not look so pretty.

**Reply by sifran-github** [reply]
do you want to do a brainstorm session about it? @dpolivaev @macmarrum and anyone else?

**Reply by sifran-github** [reply]
OK I wrote my thoughts about improving the layouts, here: https://github.com/freeplane/freeplane/discussions/1101

**Comment by sifran-github** [comment]
If this change is implemented, then please change the layout image:
![image](https://user-images.githubusercontent.com/5730355/232164175-aba81901-de3a-4023-93e9-3b0d29412ded.png)
and also for the other side.

I personally think that there can be much more possibilities of layouts.

- We can allow many starting locations where children are placed.
- We can maybe break the "layout" into individual properties, therefore allowing much more configurations.
- I have more ideas/thoughts...

I don't mind having some brainstorm about it, and contribute my thoughts, if people want to.

**Reply by macmarrum** [reply]
I'd like to learn about the diagram types you'd have in mind, like an Ishikawa diagram. 
For each, some set of childNodesLayouts might be required. I'd find it interesting to discuss how to build such diagrams in Freeplane and then whether any new childNodesLayouts are required.

**Reply by sifran-github** [reply]
I really do NOT support the idea of childNodesLayouts (defining how children should behave, in this case, their layout).
It is complicated, it will create many conflicting layouts and will result in x10 more bugs.

I support creating it by yourself. with these tips, https://github.com/freeplane/freeplane/discussions/1066#discussioncomment-5468575, it is very simple to create and maintain.

edit: i understand that "Ishikawa diagram" is a special case where the children's layout is alternating between top and bottom. I think that it is a very special case, and Freeplane should not support it as a "baseline", and if someone wants to implement it, it can be done as an add-on that will keep this format (that in my opinion, very few people will actually use).

and even with  "Ishikawa diagram" format, what if i want "2 children up, 2 children down, and so on?" or "1 up, 2 down"? i can have many ideas about how i want to have my specific maps. with the ability to define it manually, everybody can do everything quite simply. and with scripting/add-ons, rare edge cases can be achieved.

**Reply by sifran-github** [reply]
it is also endless.

theoretically i can define: i want the node layout to be X, and this children to be Y, and grandchildren should behave by alternating between left and right.

and also.. do it only if their color is blue

i mean, its endless...
should be done in scripting in my opinion.

**Reply by macmarrum** [reply]
> theoretically i can define: i want the node layout to be X, and this children to be Y, and grandchildren should behave by alternating between left and right.

It's difficult for me to discuss theoretical ideas. It's easier for me to talk about concrete needs, like an Isikawa diagram that I mentioned.

> edit: i understand that "Ishikawa diagram" is a special case where the children's layout is alternating between top and bottom. I think that it is a very special case, and Freeplane should not support it as a "baseline", and if someone wants to implement it, it can be done as an add-on that will keep this format (that in my opinion, very few people will actually use).

When you say "it can be done as an add-on", do you already have an idea in your mind how to implement it, using the current scripting API (1), or do you mean that the scripting API should be extended to allow implementing an add-on that will support e.g. Ishikawa diagram (2)?

Either way, I'd be interested in learning about more details. In particular, in case of (1), would you be able to write such an add-on? Maybe as a proof of concept? And in case of (2), what changes to the API would you envisage to support writing such an add-on?

**Reply by sifran-github** [reply]
> It's difficult for me to discuss theoretical ideas. It's easier for me to talk about concrete needs, like an Isikawa diagram that I mentioned.

OK here is an example:
![image](https://user-images.githubusercontent.com/5730355/232284002-c0097e55-27f6-4874-82e8-0dfced03b27e.png)
I think that it might be useful sometimes. And it is nice to be able to manually define which parts are "up" and which parts are "down" by changing the layout of every blue node.

> When you say "it can be done as an add-on", do you already have an idea in your mind how to implement it, using the current scripting API (1), or do you mean that the scripting API should be extended to allow implementing an add-on that will support e.g. Ishikawa diagram (2)?

If you just want to alternate between top and bottom, automatically, one by one, then... yes, i have an idea: a script that allows you to define layout1, and layout2, and then when you select the parent and run the script, it automatically changes the layouts of the children to alternate between these layouts, one by one.
If you want to automate it and automatically move nodes around whenever one of the children is being added/deleted, its another part that probably need listeners(?). I'm not sure if its possible but better add-on devs can help here.

> Either way, I'd be interested in learning about more details. In particular, in case of (1), would you be able to write such an add-on? Maybe as a proof of concept? And in case of (2), what changes to the API would you envisage to support writing such an add-on?

Am I going to build it? No. There are better add-on developers than me here + I don't need this add-on. Its easy to create manually + there is much more control this way, and ability to create "Ishikawa diagram", the diagram above and many more.

**Reply by sifran-github** [reply]
Again, you can create this "Ishikawa diagram" - just manually instead of automatic. right? the issue is just that it is done manually and not automatically, right?
or is there something that I am missing, and you cant create it at all?

**Reply by macmarrum** [reply]
> OK here is an example:
[image]
I think that it might be useful sometimes. And it is nice to be able to manually define which parts are "up" and which parts are "down" by changing the layout of every blue node.

OK. Great. BTW, it looks like a (small) Ishikawa diagram.

I am able to create such a diagram in Freeplane, using the childNodesLayouts available in v1.11.1-pre29, and to manually define which parts are "up" and which parts are "down" by changing the layout of every blue node. Is there anything you are missing?

**Reply by sifran-github** [reply]
Maybe we misunderstand each other. by "childNodesLayouts" I thought you meant a feature that doesn't exist in pre29 and you want to implement it.

Yeah I thought "childNodesLayouts" is some other feature. Nevermind then :D

**Reply by sifran-github** [reply]
regarding ideas - take a look at my posts above, in the previous post in this thread. I mention there that the location of the children starting position can be better broken into properties that will allow more layouts.

**Reply by macmarrum** [reply]
I apologize for any confusion. Indeed, **childNodesLayout** is a scripting-API term. The option in GUI is called **Children nodes layout**

![Children_nodes_layout](https://user-images.githubusercontent.com/7197131/232290455-ea420fe9-4a8d-4a08-9340-42081c13db96.png)

**Reply by sifran-github** [reply]
yeah, I thought you were talking about something completely different :) sorry.
